File: code/cvc5mvn/src/test/java/plaid/antlr/TypeVisitorTest.java
package plaid.antlr;

import org.junit.Test;
import plaid.ast.*;

import java.util.Map;
import java.util.TreeMap;

import static org.junit.Assert.assertEquals;

public class TypeVisitorTest {
    
    private static Type ast(String src){
        return Loader.toType(src);
    }
    
    /**
     * Parses client id type
     */
    @Test
    public void partyIndexType(){
        // using loader I can convert cid as AST, PartyIndexType() 
        assertEquals(new PartyIndexType(), ast("cid"));
    }

    /**
     * Parses string type
     */
    @Test
    public void stringType(){
        assertEquals(new StringType(), ast("string"));
    }

    /**
     * Parses record type
     */
    @Test
    public void recordType(){
        TreeMap<Identifier, Type> map = new TreeMap<>();
        map.put(new Identifier("s"), new StringType());
        map.put(new Identifier("i"), new PartyIndexType());
        RecordType expected = new RecordType(map);
        
        assertEquals(expected, ast("{s: string; i: cid}"));        
    }

    /**
     * parses nested record type
     */
    @Test
    public void nestedRecordType(){
        TreeMap<Identifier, Type> map = new TreeMap<>();
        map.put(new Identifier("s"), new RecordType(new TreeMap<>(Map.of(new Identifier("i"), new PartyIndexType()))));
        RecordType expected = new RecordType(map);

        assertEquals(expected, ast("{s: {i: cid}}"));
    }

}

File: code/cvc5mvn/src/test/scala/plaid/antlr/TypeVisitorTest.scala
package plaid.antlr

import org.junit.Test
import plaid.ast._
import scala.jdk.CollectionConverters._
import scala.collection.immutable.TreeMap
import org.junit.Assert.assertEquals

class TypeVisitorTest {

  private def ast(src: String): Type = Loader.toType(src)

  /** Parses client id type */
  @Test
  def partyIndexType(): Unit = {
    assertEquals(PartyIndexType(), ast("cid"))
  }

  /** Parses string type */
  @Test
  def stringType(): Unit = {
    assertEquals(StringType(), ast("string"))
  }

  /** Parses record type */
  @Test
  def recordType(): Unit = {
    val map: TreeMap[Identifier, Type] = TreeMap(
      Identifier("s") -> StringType(),
      Identifier("i") -> PartyIndexType()
    )

    val expected = RecordType(java.util.TreeMap(map.asJava))
    assertEquals(expected, ast("{s: string; i: cid}"))
  }

  /** Parses nested record type */
  @Test
  def nestedRecordType(): Unit = {
    val nestedMap: TreeMap[Identifier, Type] = TreeMap(
      Identifier("i") -> PartyIndexType()
    )
    val map: TreeMap[Identifier, Type] = TreeMap(
      Identifier("s") -> RecordType(java.util.TreeMap(nestedMap.asJava))
    )

    val expected = RecordType(java.util.TreeMap(map.asJava))
    assertEquals(expected, ast("{s: {i: cid}}"))
  }
}
